### PR TITLE
feat: recommender

### PR DESCRIPTION
**Pull request by SzymonMrzyglod**

### CODE REVIEW

File: src/locales/en-US/recommender.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  'recommender.too_long': 'Lessons that are too long cause a drop in pass rates ',
+  'recommender_info.title': 'How do we know this?',
+  'recommender_info.first_title': 'Recommender for better Educational results.',
+  'recommender_info.first_text':
+    'Wellms continually strives to improve the experience and ease of adding multimedia content. How can we be sure that subsequent exercise and lesson recommendations are worth adding? The answer lies in advanced artificial intelligence algorithms and precise calculations, which are an integral part of our system based on XGBoost, i.e. eXtreme Gradient Boosting, algorithms.',
+  'recommender_info.second_title': 'The recommender analyzes the content:',
+  'recommender_info.second_text':
+    'The Wellms Recommender analyzes the content of each exercise and lesson for complexity, comprehensibility, and consistency with other course materials. Recommending next steps or completion percentage. Machine learning algorithms analyze patterns in the data, helping to identify exercises and lessons that best suit students individual preferences and skill levels.',
+  'recommender.add_new_exercise': 'Add exercise',
+};
```
Comment (Haxikowy) [inline src/locales/en-US/recommender.ts:17]: Wouldn't it be better?
```suggestion
export default {
  recommender: {
    title: "Recommender Wellms",
    add_exercise: "Add an exercise now ",
    increase: "to increase the course pass rate ",
    add_button: "Yes, add",
    no_recommendation: "No recommendations",
    finish: "Finish this lesson now. ",
    too_long: "Lessons that are too long cause a drop in pass rates ",
    add_new_exercise: "Add exercise",
    info: {
      title: "How do we know this?",
      first_title: "Recommender for better Educational results.",
      first_text:
        "Wellms continually strives to improve the experience and ease of adding multimedia content. How can we be sure that subsequent exercise and lesson recommendations are worth adding? The answer lies in advanced artificial intelligence algorithms and precise calculations, which are an integral part of our system based on XGBoost, i.e. eXtreme Gradient Boosting, algorithms.",
      second_title: "The recommender analyzes the content:",
      second_text:
        "The Wellms Recommender analyzes the content of each exercise and lesson for complexity, comprehensibility, and consistency with other course materials. Recommending next steps or completion percentage. Machine learning algorithms analyze patterns in the data, helping to identify exercises and lessons that best suit students individual preferences and skill levels.",
    },
  },
};
```

File: src/locales/pl-PL/recommender.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  'recommender_info.title': 'Skąd to wiemy?',
+  'recommender_info.first_title': 'Rekomender dla lepszych wyników Edukacyjnych.',
+  'recommender_info.first_text':
+    'Wellms nieustannie dąży do poprawy doświadczenia i łatwości dodawania materiałów multimedialnych. Skąd możemy mieć pewność, że kolejne rekomendacje ćwiczeń i lekcji są warte dodania? Odpowiedź tkwi w zaawansowanych algorytmach sztucznej inteligencji i precyzyjnych obliczeniach, które są integralną częścią naszego systemu opartych na algorytmach XGBoost, czyli eXtreme Gradient Boosting.',
+  'recommender_info.second_title': 'Rekomender dokonuje analizy zawartości:',
+  'recommender_info.second_text':
+    'Rekomender Wellms analizuje zawartość każdego ćwiczenia i lekcji pod kątem złożoności, zrozumiałości i spójności z innymi materiałami kursu. Rekomendując kolejne kroki albo procent ukończalności. Algorytmy uczenia maszynowego analizują wzorce w danych, pomagając zidentyfikować ćwiczenia i lekcje, które najlepiej odpowiadają indywidualnym preferencjom i poziomowi zaawansowania uczniów.',
+  'recommender.add_new_exercise': 'Dodaj ćwiczenie',
+};
```
Comment (Haxikowy) [inline src/locales/pl-PL/recommender.ts:17]: Wouldn't it be better?
```suggestion
export default {
  recommender: {
    title: "Rekomender Wellms",
    add_exercise: "Dodaj teraz ćwiczenie ",
    increase: "aby podnieść zdawalność kursu o ",
    add_button: "Tak, dodaj",
    no_recommendation: "Brak rekomendacji",
    finish: "Zakończ tę lekcję teraz. ",
    too_long: "Zbyt długie lekcje powodują spadek zdawalności o ",
    add_new_exercise: "Dodaj ćwiczenie",
    info: {
      title: "Skąd to wiemy?",
      first_title: "Rekomender dla lepszych wyników Edukacyjnych.",
      first_text:
        "Wellms nieustannie dąży do poprawy doświadczenia i łatwości dodawania materiałów multimedialnych. Skąd możemy mieć pewność, że kolejne rekomendacje ćwiczeń i lekcji są warte dodania? Odpowiedź tkwi w zaawansowanych algorytmach sztucznej inteligencji i precyzyjnych obliczeniach, które są integralną częścią naszego systemu opartych na algorytmach XGBoost, czyli eXtreme Gradient Boosting.",
      second_title: "Rekomender dokonuje analizy zawartości:",
      second_text:
        "Rekomender Wellms analizuje zawartość każdego ćwiczenia i lekcji pod kątem złożoności, zrozumiałości i spójności z innymi materiałami kursu. Rekomendując kolejne kroki albo procent ukończalności. Algorytmy uczenia maszynowego analizują wzorce w danych, pomagając zidentyfikować ćwiczenia i lekcje, które najlepiej odpowiadają indywidualnym preferencjom i poziomowi zaawansowania uczniów.",
    },
  },
};
```

File: src/icons/index.tsx [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import Icon from '@ant-design/icons';
+import type { CustomIconComponentProps } from '@ant-design/icons/lib/components/Icon';
+import { RecommenderSvg } from './recommender';
+
+export const RecommenderIcon = (props: Partial<CustomIconComponentProps>) => (
+  <Icon component={RecommenderSvg} {...props} />
+);
```
Comment (Haxikowy) [inline src/icons/index.tsx:7]: Nicely done!

File: src/components/ProgramForm/ThreeColProgram/List/index.css [hunk 4]
```diff
@@ -163,3 +162,59 @@
...
+
+/* RECOMMENDER */
+.recommender {
+  position: relative;
+  display: flex;
+  gap: 8px;
+
+  margin: 8px;
+  padding: 14px 14px 14px 0;
+  background-color: #dbfddc;
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/index.css:174]: I would create custom CSS var for this. (in the recommender block, so in this class)

File: src/components/ProgramForm/ThreeColProgram/List/index.css [hunk 5]
```diff
@@ -163,3 +162,59 @@
...
+  border-radius: 12px;
+}
+.recommender__title {
+  color: #2dac34;
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/index.css:178]: I would create custom CSS var for this. (in the recommender block)

File: src/components/ProgramForm/ThreeColProgram/List/index.css [hunk 6]
```diff
@@ -163,3 +162,59 @@
...
+  font-weight: 600;
+  font-size: 12px;
+}
+.recommender__paragraph {
+  font-size: 10px;
+}
+.recommender__paragraph--span {
+  color: #2dac34;
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/index.css:186]: I would create custom CSS var for this. (in the recommender block)

File: src/components/ProgramForm/ThreeColProgram/List/index.css [hunk 7]
```diff
@@ -163,3 +162,59 @@
...
+  font-weight: 600;
+  font-size: 10px;
+  background-color: #2dac34;
+  border-color: #2dac34;
+  border-radius: 2px;
+}
+.recommender__button:hover {
+  color: white;
+  background-color: #2dac34;
+  border-color: #2dac34;
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/index.css:201]: I would create custom CSS var for this. (in the recommender block)

File: src/components/ProgramForm/ThreeColProgram/List/index.css [hunk 8]
```diff
@@ -163,3 +162,59 @@
...
+.recommender__paragraph {
+  font-size: 10px;
+}
+.recommender__paragraph--span {
+  color: #2dac34;
+  font-weight: 600;
+}
+.recommender__button {
+  padding: 5px 12px;
+  color: white;
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/index.css:191]: I would avoid `keyword colors` - create custom CSS var in recommender block (for this scope only).

File: src/components/ProgramForm/ThreeColProgram/List/index.css [hunk 9]
```diff
@@ -163,3 +162,59 @@
...
+  font-weight: 600;
+  font-size: 10px;
+  background-color: #2dac34;
+  border-color: #2dac34;
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/index.css:195]: I would create custom CSS var for this. (in the recommender block)

File: src/components/ProgramForm/ThreeColProgram/List/index.css [hunk 10]
```diff
@@ -163,3 +162,59 @@
...
+  border-radius: 2px;
+}
+.recommender__button:hover {
+  color: white;
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/index.css:199]: I would avoid `keyword colors` - create custom CSS var in recommender block (for this scope only).

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    (lesson_id: number, topic_type: TopicType) => {
+      if (addNewTopic) {
+        const newTopic = addNewTopic(lesson_id, topic_type);
+        history.push(`/courses/list/${courseId}/program/?topic=${newTopic.id}`);
+      }
+    },
+    [addNewTopic],
+  );
+
+  const getTopicType = (topic: API.RecommenderTopicType) => (topic.includes('H5P') ? 'H5P' : topic);
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx:32]: This should be outside of component scope - it's pure function. I would also add return type.

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        );
+      default:
+        return <FormattedMessage id="no_recommendation" />;
+    }
+  };
+
+  useEffect(() => {
+    getCourseRecommender(courseId).then((res) => res.success && setCourse(res.data));
+    getExerciseRecommender(lessonId).then((res) => res.success && setExcercise(res.data));
+  }, [courseId, lessonId]);
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx:57]: This should be separated to two different `useEffect` hooks. For readability I would even create custom hook/context.

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+                onTopicCreate(lessonId, TopicType[getTopicType(excercise?.topic_type)])
+              }
+              recommendedData={excercise}
+            />
+          )
+        );
+      default:
+        return <FormattedMessage id="no_recommendation" />;
+    }
+  };
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx:52]: I'm not fan of fn like this. I would simply move it to JSX.

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderTopicSelector.tsx [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+      switch (type) {
+        case TopicType.GiftQuiz:
+          return !(initialState?.packages && initialState?.packages?.[PACKAGES.TopicTypeGift]);
+        case TopicType.SCORM:
+          return !(initialState?.packages && initialState?.packages?.[PACKAGES.Scorm]);
+        case TopicType.Project:
+          return !(initialState?.packages && initialState?.packages?.[PACKAGES.TopicTypeProject]);
+        default:
+          return !(initialState?.packages && initialState?.packages?.[PACKAGES.TopicTypes]);
+      }
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderTopicSelector.tsx:60]: You can utilize `createHavePackageInstalled` factory here to make code more readable.

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderTopicSelector.tsx [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+import '../types.css';
+import { TopicType } from '@/services/escola-lms/enums';
+import { FormattedMessage, useModel } from 'umi';
+import { topicTypeToSettingName } from '@/pages/Settings/global';
+
+export const RecommenderTopicSelector: React.FC<{
+  onSelected: (type: TopicType) => void;
+  positionsToHide?: (TopicType | 'lesson')[];
+}> = ({ onSelected, positionsToHide }) => {
+  const [open, setOpen] = useState<boolean>(false);
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderTopicSelector.tsx:28]:
```suggestion
  const [open, setOpen] = useState(false);
```

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderTopicSelector.tsx [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+      >
+        <FormattedMessage id="recommender.add_new_exercise" />
+      </Button>
+
+      <div
+        className={`${'topic-types-selector__recommender'}
+          ${open ? 'topic-types-selector__recommender--active' : ''}`}
+      >
+        {!positionsToHide?.includes(TopicType.RichText) &&
+          !topicTypeIsDisabled(TopicType.RichText) && (
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderTopicSelector.tsx:85]: I see room for improvements here. I would probably move it to custom fn e.g.

```ts
const isTopicVisible = (type: TopicType): boolean => !positionsToHide?.includes(type) && !topicTypeIsDisabled(type)
```

Also I would consider mapping the topic types here, as You are repeating a lot of code.

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderInfo.tsx [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+    <p className="recommender__popover-text">
+      <FormattedMessage id="recommender_info.second_text" />
+    </p>
+  </div>
+);
+
+export const RecommenderInfo: React.FC = () => (
+  <Popover
+    placement="bottomLeft"
+    content={popoverContent}
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderInfo.tsx:28]:
```suggestion
const PopoverContent = () => (
  <div className="recommender__popover-wrapper">
    <h3 className="recommender__popover-title">
      <FormattedMessage id="recommender_info.title" />
    </h3>
    <strong className="recommender__popover-text">
      <FormattedMessage id="recommender_info.first_title" />
    </strong>
    <p className="recommender__popover-text">
      <FormattedMessage id="recommender_info.first_text" />
    </p>
    <strong className="recommender__popover-text">
      <FormattedMessage id="recommender_info.second_title" />
    </strong>
    <p className="recommender__popover-text">
      <FormattedMessage id="recommender_info.second_text" />
    </p>
  </div>
);

export const RecommenderInfo: React.FC = () => (
  <Popover
    placement="bottomLeft"
    content={<PopoverContent />}
```

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/ExerciseRecommender.tsx [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+  recommendedData: API.RecommenderExercise;
+  createTopic: () => void;
+}
+
+export const ExerciseRecommender: React.FC<ExerciseRecommenderProps> = ({
+  recommendedData,
+  createTopic,
+}) => {
+  const createText = (topic: API.RecommenderTopicType) =>
+    topic?.includes('.') && topic?.split('.').join(' ');
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/ExerciseRecommender.tsx:15]: Move it outside of component scope, it's pure function. I suggest adding return type here.

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/ExerciseRecommender.tsx [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    <>
+      <Paragraph className="recommender__paragraph">
+        <FormattedMessage id="recommender.add_exercise" />
+        <span className="recommender__paragraph--span">{`${createText(
+          recommendedData?.topic_type,
+        )},`}</span>
+        <br />
+        <FormattedMessage id="recommender.increase" />
+        <span className="recommender__paragraph--span">
+          {`^ ${Math.floor(recommendedData.probability * 100)}%`}
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/ExerciseRecommender.tsx:27]: I see that this code is repeated. I would suggest moving it to custom fn e.g.:
```ts
const formatRecommenderProbability = (probability: number): number => Math.floor(probability * 100)
```

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/CourseRecommender.tsx [hunk 20]
```diff
@@ -0,0 +1,16 @@
...
+  recommendedData: API.Recommender;
+}
+
+export const CourseRecommender: React.FC<CourseRecommenderProps> = ({ recommendedData }) => (
+  <Paragraph className="recommender__paragraph">
+    <strong>
+      <FormattedMessage id="recommender.finish" />
+    </strong>
+    <FormattedMessage id="recommender.too_long" />
+    <strong>{`${Math.floor(recommendedData.probability * 100)}%`}</strong>
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/CourseRecommender.tsx:14]: I see that this code is repeated. I would suggest moving it to custom fn e.g.:
```ts
const formatRecommenderProbability = (probability: number): number => Math.floor(probability * 100)
```

File: src/locales/en-US/recommender.ts (hunk 1 above)
Comment (SzymonMrzyglod) [reply src/locales/en-US/recommender.ts:17]: We do it this way everywhere:
Zrzut ekranu 2023-11-14 o 07 41 30

File: src/locales/pl-PL/recommender.ts (hunk 2 above)
Comment (SzymonMrzyglod) [reply src/locales/pl-PL/recommender.ts:17]: We do it this way everywhere

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx (hunk 13 above)
Comment (SzymonMrzyglod) [reply src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx:52]: In my opinion, using a switch case is more clear. I think we can leave it as it is now.

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderTopicSelector.tsx (hunk 14 above)
Comment (SzymonMrzyglod) [reply src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderTopicSelector.tsx:60]: I don't need it here, it's just part of the recommender.

Comment (Haxikowy) [reply src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderTopicSelector.tsx:113]: Bump

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderInfo.tsx (hunk 17 above)
Comment (Haxikowy) [reply src/components/ProgramForm/ThreeColProgram/List/Recommender/RecommenderInfo.tsx]: Bump, create a custom component from it.

File: src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+          )
+        );
+      default:
+        return <FormattedMessage id="recommender.no_recommendation" />;
+    }
+  };
+
+  const showLoading = (type: RecommenderType) =>
+    (type === RecommenderType.Exercise && !exercise) ||
+    (type === RecommenderType.Course && !course);
```
Comment (Haxikowy) [inline src/components/ProgramForm/ThreeColProgram/List/Recommender/index.tsx:56]:
```suggestion
  const isLoading = (recommenderType === RecommenderType.Exercise && !exercise) ||
    (recommenderType === RecommenderType.Course && !course);
```

**Review by Haxikowy — APPROVED** [review]

**Pull request merged**